pkg/api/controller.go

```diff
@@ -1238,20 +1240,33 @@ func (c *Controller) CreateRepository(w http.ResponseWriter, r *http.Request, bo
 	writeResponse(w, http.StatusCreated, response)
 }
 
-func ensureStorageNamespace(ctx context.Context, adapter block.Adapter, storageNamespace string) error {
+var errorNonEmptyStorageNamespace = errors.New("non-empty storage namespace")
+
+func (c *Controller) ensureStorageNamespace(ctx context.Context, storageNamespace string) error {
 	const (
 		dummyKey  = "dummy"
 		dummyData = "this is dummy data - created by lakeFS in order to check accessibility"
 	)
 
 	obj := block.ObjectPointer{StorageNamespace: storageNamespace, Identifier: dummyKey}
 	objLen := int64(len(dummyData))
-	err := adapter.Put(ctx, obj, objLen, strings.NewReader(dummyData), block.PutOpts{})
+	err := c.BlockAdapter.Put(ctx, obj, objLen, strings.NewReader(dummyData), block.PutOpts{})
 	if err != nil {
 		return err
 	}
-	_, err = adapter.Get(ctx, obj, objLen)
-	return err
+	if _, err = c.BlockAdapter.Get(ctx, obj, objLen); err != nil {
+		return err
+	}
+
+	// if a single file exists under the lakeFS prefix in the storage namespace
```

Comment:
AFAICS this allows multiple repositories to use the same storage namespace, as long as I don't use them first.  (So it is also a race if I create both repos and then upload a file to each.)

Why not just check for existence of the "dummy" file?  (It would _still_ be a race, but at least it would prevent this case.)

Thinking about it, one safe way to do this without a full database would be to keep a Graveler object with all storage namespaces.  Then you could verify-nonoverlap-and-add, and use a commit-like operation to prevent races.  Maybe you could have a hidden "global lakeFS" repo and keep a file named for each storage namespace.

Anyway, the current code is racy -- and since the time between creating a repo and using it is **unbounded**, there is no protection against this race.

Reply:
+1 for using the dummy file.
I think we can live with the race- but if we really want to solve it we can implement a conditional put in the block adapter (like If-None-Match in S3).

Note that using a "global lakeFS" repository will not solve the race for multiple lakeFS installations (to me it's not really a problem but just noting).

Reply:
Yeah, I missed Oz remarks of the requirements in #3077 - we need to tighten it some more.
The requirements states cross-installations enforcement, so a graveler object won't do the trick here.

I avoided checking the `dummy` object because I feel less comfortable treating it as a contract between repositories. It's there because of a one-time check, no one ever checks it the second time, and it may very well be deleted without us knowing (GC, lifecycle policies, etc..). But I seem to be the only one who believes it's fragile, so I'll do this check..

Reply:
For example, now that we delete a repo we also need to delete that `dummy` file.